Bloom.md

```diff
@@ -0,0 +1,341 @@
+| 100 | 2535301200456458804968143894986       |
+
+Practically, we can expect scaling to stop well before the 50 range. With the default capacity of 100,000 and the default
+expansion rate of 2, the total items inserted (across all filters) would exceed the 64-bit unsigned integer limit before
+reaching 50 filters.
+
+
+### Choice of Bloom Filter Library
+
+We evaluated the following libraries:
+* https://crates.io/crates/bloomfilter
+* https://crates.io/crates/growable-bloom-filter
+* https://crates.io/crates/fastbloom
+
+Here are some important factors when evaluating libraries:
+* Is it actively maintained?
+* Has there been any breaking changes in serialization and deserialization of bloom filters over version updates?
+* Performance
+* Popularity
+
+https://crates.io/crates/bloomfilter was chosen as it provides all the required APIs for controlling a bloom filter.
+This library has also been stable through the releases.
+
+Certain libraries (e.g. fastbloom) are no longer compatible with older versions (after updates) resuling in
+serialization / deserialization incompatibility.
+
+growable-bloom-filter is a bloom filter library that implements a scalable bloom filter. However, it does not allow
+control over when the scaling occurs and does not expose APIs to check the number of filters that it has scaled out to.
+This crate also has had a breaking change across versions and bloom filters of the older version are not loadable on the
+newer versions.
+
+
+### Large BloomFilter objects
+
+Create and Delete operations on large bloom filters take longer durations and will block the main thread for this duration. 
+Because of this, the following operations will be handled differently:
+* defrag callback: If the memory used by any bloom filter within the bloom object is greater than X MB (`bloom_large_item_threshold`
+    constant), we will skip defrag operations on this bloom object. Otherwise, we will defrag the bloom object in iterations
+    for each bloom filter in the bloom object.
+* free_effort callback: This callback decides the free effort for the bloom object. If it is greater than X MB
+    (`bloom_large_item_threshold` constant), we will return 0 to use async free on the bloom object.
+* create operations (BF.ADD/MADD/INSERT/RESERVE): We can compute the estimated size of the bloom filter that will be created as
+    a result of the operation and we can consider rejecting requests in case of unavailable memory. Scalable Bloom filters
+    will grow in used memory after creation - but only as a result of a BF.ADD, BF.MADD, or BF.INSERT operation and we can reject
+    these requests if we identify that it will result in a scale out operation that requires more memory than what is available.
+
+
+## Specification
+
+### Bloom Filter Command API
+
+The following Bloom Filter commands are supported and their API syntax is compatible with ReBloom:
+BF.ADD <key> <item>
+BF.EXISTS <key> <item>
+BF.MADD <key> <item> [<item> ...]
```

Comment:
do we really need MADD? Can't ADD itself support this?

Reply:
There used to be a lot of discussion around variadic vs non-variadic commands. Today the return type of BF.ADD is an integer reply in RedisBloom, whether or not the item was inserted. BF.MADD is an array of integers. We could alter the return type of BF.ADD to be an array (but then it wouldn't be API compatible with RedisBloom) but also then makes the API kind of weird since you are given an array of one element in many cases.

Reply:
Can't we be smart about it? If there is only 1 return value - do non-array otherwise array. If clients want to expose it as array only, they cast the single object to array. If this is already a closed topic, we can skip rehashing the discussion here.

Reply:
No, clients hate having variadic return types. Then you have to write custom client code to handle.

> If this is already a closed topic, we can skip rehashing the discussion here.

We used to say the only thing consistent about Redis is that it was inconsistent. I'm just trying to pass on context for decisions, feel free to argue with them :)